File: frontend/src/services/restart.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

/**
* Interface that represents the restart state of players
*/
export interface PlayerRestartStatus {
  gameID: string;
  nicknameB: string;
  nicknameW: string;
  restartB: boolean;
  restartW: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class RestartService {

  constructor(private http: HttpClient) {}

  /**
  * Gets the current restart state for a specific game
  * @param gameId - Game ID
  * @returns Observable with the restart state
  */
  getRestartStatus(gameId: string): Observable<PlayerRestartStatus> {
    return this.http.get<PlayerRestartStatus>(`/api/restartStatus/${gameId}/`);
  }

  /**
  * Updates a player's restart status
  * @param status - Object with the updated restart status
  * @returns Observable with the response
  */
  updateRestartStatus(status: PlayerRestartStatus): Observable<any> {
    return this.http.post(`/api/restartStatus/${status.gameID}`, status);
  }

  /**
  * Resets the game when both players have accepted
  * @param gameId - Game ID to reset
  * @returns Observable with the response
  */
  resetGame(gameId: string): Observable<any> {
    return this.http.post(`/api/games/${gameId}/reset`, {});
  }

  /**
  * Utility to determine if the current player is white
  * @param status - Restart status
  * @param nickname - Nickname of the current player
  * @returns true if the player is white
  */
  isWhitePlayer(status: PlayerRestartStatus, nickname: string | null): boolean {
    return nickname === status.nicknameW;
  }

  /**
  * Utility to determine if the current player is black
  * @param status - Restart status
  * @param nickname - Nickname of the current player
  * @returns true if the player is black
  */
  isBlackPlayer(status: PlayerRestartStatus, nickname: string | null): boolean {
    return nickname === status.nicknameB;
  }

  /**
  * Updates the white player's restart status
  * @param status - Current restart status
  * @param wantsRestart - New value
  * @returns New state object with the updated value
  */
  setWhitePlayerRestart(status: PlayerRestartStatus, wantsRestart: boolean): PlayerRestartStatus {
    return {
      ...status,
      restartW: wantsRestart
    };
  }

  /**
  * Updates the black player's restart status
  * @param status - Current restart status
  * @param wantsRestart - New value
  * @returns New state object with the updated value
  */
  setBlackPlayerRestart(status: PlayerRestartStatus, wantsRestart: boolean): PlayerRestartStatus {
    return {
      ...status,
      restartB: wantsRestart
    };
  }

  /**
  * Check if both players have requested a restart
  * @param status - Restart status
  * @returns true if both players want to restart
  */
  bothPlayersWantRestart(status: PlayerRestartStatus,): boolean {
    return status.restartB && status.restartW;
  }

  resetPlayerRestart(gameId:string) {
    return this.http.post<PlayerRestartStatus>(`/api/restartStatus/${gameId}/restart`, {});
  }
}

File: backend/api-server/src/main/java/org/onlinecheckers/apiserver/services/RestartService.java
package org.onlinecheckers.apiserver.services;

import org.onlinecheckers.apiserver.model.dtos.PlayerRestartDto;
import org.onlinecheckers.apiserver.repositories.PlayerRestartRepository;
import org.onlinecheckers.apiserver.exceptions.SessionGameNotFoundException;

import org.springframework.stereotype.Service;
import org.springframework.beans.factory.annotation.Autowired;

@Service
public class RestartService {

    @Autowired
    private PlayerRestartRepository playerRestartRepository;

    /**
     * Get the restart status for a specific game
     * @param gameId - ID of the game
     * @return PlayerRestartDto with current restart status
     * @throws SessionGameNotFoundException if game not found
     */
    public PlayerRestartDto getRestartStatus(String gameId) {
        return playerRestartRepository.findById(gameId)
                .orElseThrow(() -> new SessionGameNotFoundException(gameId));
    }

    /**
     * Update the restart status for a game
     * @param status - Updated restart status
     * @return Saved PlayerRestartDto
     */
    public PlayerRestartDto updateRestartStatus(PlayerRestartDto status) {
        return playerRestartRepository.save(status);
    }

    /**
     * Reset the restart status for both players (set both flags to false)
     * @param gameId - ID of the game
     * @return Updated PlayerRestartDto with both restart flags set to false
     * @throws SessionGameNotFoundException if game not found
     */
    public PlayerRestartDto resetPlayerRestart(String gameId) {
        PlayerRestartDto restart = getRestartStatus(gameId);
        restart.setRestartB(false);
        restart.setRestartW(false);
        return playerRestartRepository.save(restart);
    }

    /**
     * Check if both players want to restart
     * @param status - Current restart status
     * @return true if both players have requested restart
     */
    public boolean bothPlayersWantRestart(PlayerRestartDto status) {
        return status.isRestartB() && status.isRestartW();
    }

    /**
     * Create initial restart status for a new game
     * @param gameId - ID of the game
     * @param nicknameW - White player nickname
     * @param nicknameB - Black player nickname
     * @return Created PlayerRestartDto
     */
    public PlayerRestartDto createInitialRestartStatus(String gameId, String nicknameW, String nicknameB) {
        PlayerRestartDto restartStatus = new PlayerRestartDto();
        restartStatus.setGameID(gameId);
        restartStatus.setNicknameW(nicknameW);
        restartStatus.setNicknameB(nicknameB);
        restartStatus.setRestartW(false);
        restartStatus.setRestartB(false);
        return playerRestartRepository.save(restartStatus);
    }

    /**
     * Update restart status for a specific player
     * @param gameId - ID of the game
     * @param playerNickname - Nickname of the player
     * @param wantsRestart - Whether the player wants to restart
     * @return Updated PlayerRestartDto
     */
    public PlayerRestartDto updatePlayerRestartStatus(String gameId, String playerNickname, boolean wantsRestart) {
        PlayerRestartDto status = getRestartStatus(gameId);
        
        if (playerNickname.equals(status.getNicknameW())) {
            status.setRestartW(wantsRestart);
        } else if (playerNickname.equals(status.getNicknameB())) {
            status.setRestartB(wantsRestart);
        } else {
            throw new IllegalArgumentException("Player " + playerNickname + " not found in game " + gameId);
        }
        
        return playerRestartRepository.save(status);
    }

    /**
     * Check if a specific player has requested restart
     * @param status - Current restart status
     * @param playerNickname - Nickname of the player to check
     * @return true if the player has requested restart
     */
    public boolean hasPlayerRequestedRestart(PlayerRestartDto status, String playerNickname) {
        if (playerNickname.equals(status.getNicknameW())) {
            return status.isRestartW();
        } else if (playerNickname.equals(status.getNicknameB())) {
            return status.isRestartB();
        }
        return false;
    }
}